DatadogRUM/Sources/SDKMetrics/SessionEndedMetricController.swift

```diff
@@ -0,0 +1,88 @@
+    private let telemetry: Telemetry
+
+    /// Initializes a new instance of the metric controller.
+    /// - Parameter telemetry: The telemetry endpoint used for sending metrics.
+    init(telemetry: Telemetry) {
+        self.telemetry = telemetry
+    }
+
+    /// Starts a new metric for a given session.
+    /// - Parameters:
+    ///   - sessionID: The ID of the session to track.
+    ///   - precondition: The precondition that led to starting this session.
+    ///   - context: The SDK context at the moment of starting this session.
+    /// - Returns: The newly created `SessionEndedMetric` instance.
+    func startMetric(sessionID: RUMUUID, precondition: RUMSessionPrecondition?, context: DatadogContext) {
+        guard sessionID != RUMUUID.nullUUID else {
+            return // do not track metric when session is not sampled
+        }
+        let metric = SessionEndedMetric(sessionID: sessionID, precondition: precondition, context: context)
+        metricsBySessionID[sessionID] = metric
+        pendingSessionIDs.append(sessionID)
+    }
+
+    /// Tracks the view event that occurred during the session.
+    /// - Parameters:
+    ///   - view: the view event to track
+    ///   - sessionID: session ID to track this view in (pass `nil` to track it for the last started session)
+    func track(view: RUMViewEvent, in sessionID: RUMUUID?) {
+        updateMetric(for: sessionID) { try $0?.track(view: view) }
+    }
+
+    /// Tracks the kind of SDK error that occurred during the session.
+    /// - Parameters:
+    ///   - sdkErrorKind: the kind of SDK error to track
+    ///   - sessionID: session ID to track this error in (pass `nil` to track it for the last started session)
+    func track(sdkErrorKind: String, in sessionID: RUMUUID?) {
+        updateMetric(for: sessionID) { $0?.track(sdkErrorKind: sdkErrorKind) }
+    }
+
+    /// Signals that the session was stopped with `stopSession()` API.
+    /// - Parameter sessionID: session ID to mark as stopped (pass `nil` to track it for the last started session)
+    func trackWasStopped(sessionID: RUMUUID?) {
+        updateMetric(for: sessionID) { $0?.trackWasStopped() }
+    }
+
+    /// Ends the metric for a given session, sending it to telemetry and removing it from pending metrics.
+    /// - Parameter sessionID: The ID of the session to end the metric for.
+    func endMetric(sessionID: RUMUUID) {
+        guard let metric = metricsBySessionID[sessionID] else {
+            return
+        }
+        telemetry.metric(name: SessionEndedMetric.Constants.name, attributes: metric.asMetricAttributes())
+        metricsBySessionID[sessionID] = nil
+        pendingSessionIDs.removeAll(where: { $0 == sessionID }) // O(n), but "ending the metric" is very rare event
+    }
+
+    private func updateMetric(for sessionID: RUMUUID?, _ mutation: (inout SessionEndedMetric?) throws -> Void) {
+        guard let sessionID = (sessionID ?? pendingSessionIDs.last) else {
```

Comment:
@ambushwork The `pendingSessionIDs` is more a history of pending sessions. It is true that we query only for the last element, but it might change from one query to another if `endMetric(sessionID:)` is called in between (we remove the ID from array as part of that call). If using single property to store the "last ID", we won't know how to change it after `endMetric(sessionID:)` as we wouldn't know the previous ID.

Also, even though sessions are expected to end in the same order as they were started, some iOS telemetry shows rare but existing consistency problems in session management. For that reason, we prefer to not introduce too many assumptions in "Session Ended" metric, which is another reason why we track the whole array of IDs rather than only the "latest" one.

Reply:
Thanks, quite clear for me now !